### DISCUSSION TITLE
bash: __bp_install: command not found

### DISCUSSION THREAD

**Discussion by azzamsa** [wezterm/wezterm#4356] [Q&A] [answered]
### What Operating System(s) are you running on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

GNOME 

### WezTerm version

 20230712-072601-f4abf8fd

### Ask your question!

Hi.

I get this error in Wezterm `bash: __bp_install: command not found` , but not in gnome-terminal.

```
~
🦄 bash
bash: __bp_install: command not found
(bash) >
exit
```

Maybe related to https://github.com/wez/wezterm/blob/845fa5d4aef9924eb591019c054d41542306af5d/assets/shell-integration/wezterm.sh#L309

Thanks for Wezterm! 🏅

**Comment by vollink** [comment]
Hi,

Something has set __bp_install as part of the environment variable `PROMPT_COMMAND`, but somehow did not source the `wezterm.sh` shell integration script, or "lost track" of its functions.  It is possible, perhaps, that you or a startup file startup exec'd a second bash at some point (leaving the environment variables, but clearing the loaded functions).  That's the only way I can recreate this in the way you describe it happening (but I'm just another user).

**Reply by azzamsa** [reply]
>  that you or a startup file startup exec'd a second bash at some point

Yes, it is.

- https://github.com/azzamsa/dotfiles/blob/b871ca098ba76db6bc040936e153633027628359/bash/.bashrc#L67

```bash
#
# Set .bashrc to drop into Fish
#

# Drop to Fish automatically from Bash. Otherwise, we'll have to type 'fish' every time outside `zellij`.
# https://wiki.archlinux.org/title/Fish#Setting_fish_as_interactive_shell_only
if [[ $(ps --no-header --pid=$PPID --format=comm) != "fish" && -z ${BASH_EXECUTION_STRING} ]]
then
	shopt -q login_shell && LOGIN_OPTION='--login' || LOGIN_OPTION=''
	exec fish $LOGIN_OPTION
fi
```

Do you have any workaround to get rid of this error message?

**Reply by wez** [reply]
could you `chsh` your shell to `fish` instead of `bash`? That would simplify your shell setup.

**Reply by azzamsa** [reply]
Yes, I used to change my shell using `chsh` all the time. However, nowadays, I want to keep my default shell intact for compatibility reasons. I've noticed that some login managers, Emacs packages, and apps crash or don't work correctly because they assume everyone's default shell is POSIX-compliant.

**Comment by vollink** [comment, marked as answer]
To put this another way, bash sourcing wezterm.sh is both making the `__bp_install` function available to the *then running* `bash` shell, but also adding it to the exported environment variable PROMPT_COMMAND.  On a program spawn (of any kind), "exported" environment variables are preserved, but functions are not preserved (`exec` functions have no mechanism for finding and preserving shell functions).

Since your startup for .bashrc is jumping to fish, then REMOVE the `bash` shell integration, since you are not planning on using bash (unless it's a subshell spawn under another program anyway).  This way, the `bash` integration doesn't source the wezterm.sh where the functions are created, nor does it add the `__bp_install` function to the PROMPT_COMMAND environment variable.
Then do the shell integration ENTIRELY within `fish` startup files (which I won't cover since I'm not a fish user).

**Reply by azzamsa** [reply]
Based on your comment, I now have a clean solution.

```diff
modified   bash/.bashrc
-
-#
-# Set .bashrc to drop into Fish
-#
-
-# Drop to Fish automatically from Bash. Otherwise, we'll have to type 'fish' every time outside `zellij`.
-# https://wiki.archlinux.org/title/Fish#Setting_fish_as_interactive_shell_only
-if [[ $(ps --no-header --pid=$PPID --format=comm) != "fish" && -z ${BASH_EXECUTION_STRING} ]]
-then
-	shopt -q login_shell && LOGIN_OPTION='--login' || LOGIN_OPTION=''
-	exec fish $LOGIN_OPTION
-fi

modified   wezterm/.config/wezterm/wezterm.lua 
 return {
+    default_prog = { 'zellij' },
     default_cursor_style = "BlinkingBar",
```

I removed the `fish` hook in my bash config and set `zellij` as the default program in Wezterm. My default shell in `zellij` is `fish`. So, the problem is solved!

Now, I get the bonus of not having to type `zellij` every time I open `wezterm`.
